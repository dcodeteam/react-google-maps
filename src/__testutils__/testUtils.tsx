export function getClassMockInstance(clazz: unknown) {
  return (clazz as jest.Mock).mock.results[0].value;
}

export function getFnMock<T>(fn: T): jest.Mock<T> {
  return (fn as unknown) as jest.Mock<T>;
}

export function forEachEvent<P>(
  events: unknown,
  fn: (key: keyof P, event: string) => void,
) {
  const eventsObject = events as { [key: string]: string };
  const keys = Object.keys(eventsObject) as Array<keyof P>;

  keys.forEach(key => {
    fn(key, eventsObject[String(key)]);
  });
}

export function createMockHandlers<P>(events: unknown): Partial<P> {
  const props: Partial<P> = {};

  forEachEvent<P>(events, key => {
    // eslint-disable-next-line typescript/no-explicit-any
    props[key] = jest.fn() as any;
  });

  return props;
}

export function emitEvent(
  instance: unknown,
  event: string,
  ...args: unknown[]
): void {
  (instance as NodeJS.EventEmitter).emit(event, ...args);
}
