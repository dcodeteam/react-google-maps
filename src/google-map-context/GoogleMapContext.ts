import { createContext } from "react";

export interface GoogleMapContext {
  readonly map?: google.maps.Map;
  readonly maps?: typeof google.maps;
}

// eslint-disable-next-line typescript/no-explicit-any
export const context = createContext<GoogleMapContext>({});

export const GoogleMapContextProvider = context.Provider;
export const GoogleMapContextConsumer = context.Consumer;
