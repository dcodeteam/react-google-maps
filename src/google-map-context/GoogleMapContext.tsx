import * as React from "react";

export interface GoogleMapContext {
  readonly map: google.maps.Map;
  readonly maps: typeof google.maps;
}

const { Provider, Consumer } = React.createContext<Partial<GoogleMapContext>>(
  {},
);

export interface GoogleMapContextProviderProps {
  value: GoogleMapContext;
  children: React.ReactNode;
}

export function GoogleMapContextProvider({
  value,
  children,
}: GoogleMapContextProviderProps) {
  return <Provider value={value}>{children}</Provider>;
}

export interface GoogleMapContextConsumerProps {
  children: (ctx: GoogleMapContext) => React.ReactNode;
}

export function GoogleMapContextConsumer({
  children,
}: GoogleMapContextConsumerProps) {
  return (
    <Consumer>
      {ctx =>
        !ctx.map || !ctx.maps ? null : children(ctx as GoogleMapContext)
      }
    </Consumer>
  );
}
