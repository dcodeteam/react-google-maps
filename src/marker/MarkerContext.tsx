import * as React from "react";

export interface MarkerContext {
  marker?: google.maps.Marker;
}

const { Consumer, Provider } = React.createContext<MarkerContext>({});

export const MarkerContextConsumer = Consumer;
export const MarkerContextProvider = Provider;
