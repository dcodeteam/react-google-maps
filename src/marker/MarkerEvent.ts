import { GenericEvent } from "../internal/GenericEvent";

export class MarkerEvent {
  public static readonly onClick = GenericEvent.onClick;

  public static readonly onDoubleClick = GenericEvent.onDoubleClick;

  public static readonly onRightClick = GenericEvent.onRightClick;

  public static readonly onMouseOut = GenericEvent.onMouseOut;

  public static readonly onMouseOver = GenericEvent.onMouseOver;

  public static readonly onMouseDown = GenericEvent.onMouseDown;

  public static readonly onMouseUp = GenericEvent.onMouseUp;

  public static readonly onDrag = GenericEvent.onDrag;

  public static readonly onDragStart = GenericEvent.onDragStart;

  public static readonly onDragEnd = GenericEvent.onDragEnd;

  public static readonly onPositionChanged = "position_changed";
}
