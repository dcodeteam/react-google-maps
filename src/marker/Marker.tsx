import * as React from "react";

import { GoogleMapComponent } from "../google-map-component/GoogleMapComponent";
import { GoogleMapContext } from "../google-map-context/GoogleMapContext";
import {
  createHandlerProxy,
  forEachEvent,
  pickChangedProps,
} from "../internal/PropsUtils";
import { MarkerContext, MarkerContextProvider } from "./MarkerContext";
import { MarkerEvent } from "./MarkerEvent";

interface EventProps {
  /**
   * This handlers is called when the marker icon was clicked.
   */
  onClick?: () => void;
  /**
   * This handlers is called when the marker icon was double clicked.
   */
  onDoubleClick?: () => void;
  /**
   * This handlers is called when the marker icon was clicked.
   */
  onRightClick?: () => void;

  /**
   * This handlers is called for a mouse down on the marker.
   */
  onMouseDown?: () => void;
  /**
   * This handlers is called when the mouse leaves the area of the marker icon.
   */
  onMouseOut?: () => void;
  /**
   * This handlers is called when the mouse enters the area of the marker icon.
   */
  onMouseOver?: () => void;
  /**
   * This handlers is called when for a mouse up on the marker.
   */
  onMouseUp?: () => void;

  /**
   * This handlers is called when the marker icon was clicked.
   */
  onDrag?: () => void;
  /**
   * This handlers is called when the marker icon was clicked.
   */
  onDragStart?: () => void;
  /**
   * This handlers is called when the marker icon was clicked.
   */
  onDragEnd?: () => void;

  /**
   * This handlers is called when the marker `position` property changes.
   */
  onPositionChanged?: () => void;
}

export interface MarkerProps extends EventProps {
  /**
   * Marker position.
   */
  position: google.maps.LatLngLiteral;

  /**
   * Rollover text.
   */
  title?: string;

  /**
   * If `true`, the marker is visible.
   */
  visible?: boolean;

  /**
   * If `true`, the marker receives mouse and touch events.
   */
  clickable?: boolean;

  /**
   * If `true`, the marker can be dragged.
   */
  draggable?: boolean;

  //
  // Style
  //

  /**
   * Which animation to play when marker is added to a map.
   */
  animation?: "BOUNCE" | "DROP";

  /**
   * Mouse cursor to show on hover.
   */
  cursor?: string;

  /**
   * Icon for the foreground.
   */
  icon?: string | React.ReactElement<object>;

  /**
   * Adds a label to the marker.
   *
   * The label can either be a `string`, or a [google.maps.MarkerLabel](https://developers.google.com/maps/documentation/javascript/3.exp/reference#MarkerLabel) object.
   */
  label?: string | google.maps.MarkerLabel;

  /**
   * The marker's opacity between 0.0 and 1.0.
   */
  opacity?: number;

  /**
   * Optimization renders many markers as a single static element.
   *
   * Disable optimized rendering for animated GIFs or PNGs,
   * or when each marker must be rendered as a separate DOM element
   * (advanced usage only).
   */
  optimized?: boolean;

  /**
   * Image map region definition used for drag/click.
   *
   * See also: [google.maps.MarkerShape](https://developers.google.com/maps/documentation/javascript/3.exp/reference#MarkerShape)
   */
  shape?: google.maps.MarkerShape;

  /**
   * All markers are displayed on the map in order of their zIndex,
   * with higher values displaying in front of markers with lower values.
   * By default, markers are displayed according to their vertical position on screen,
   * with lower markers appearing in front of markers further up the screen.
   */
  zIndex?: number;
}

function createOptions(
  maps: typeof google.maps,
  {
    position,
    title,
    visible,
    clickable,
    draggable,
    cursor,
    label,
    opacity,
    optimized,
    shape,
    zIndex,

    icon,
    animation,
  }: MarkerProps,
): google.maps.MarkerOptions {
  return {
    position,
    title,
    visible,
    clickable,
    draggable,
    cursor,
    label,
    opacity,
    optimized,
    shape,
    zIndex,

    icon: typeof icon === "string" ? icon : undefined,
    animation: animation && maps.Animation[animation],
  };
}

interface State {
  ctx: MarkerContext;
  // eslint-disable-next-line typescript/no-use-before-define
  marker: google.maps.Marker;
}

export class Marker extends React.Component<MarkerProps> {
  public render() {
    const { icon } = this.props;

    return (
      <GoogleMapComponent
        createInitialState={({ maps }: GoogleMapContext): State => {
          const marker = new maps.Marker();

          return { marker, ctx: { marker } };
        }}
        createOptions={({
          maps,
        }: GoogleMapContext): google.maps.MarkerOptions =>
          createOptions(maps, this.props)
        }
        didMount={({ map, options, state: { marker } }) => {
          marker.setMap(map);
          marker.setOptions(options);

          forEachEvent(MarkerEvent, (key, event) => {
            // eslint-disable-next-line react/destructuring-assignment
            const handler = createHandlerProxy(() => this.props[key]);

            if (event === MarkerEvent.onDragEnd) {
              marker.addListener(event, e => {
                marker.setPosition(options.position);

                handler(e);
              });
            } else {
              marker.addListener(event, handler);
            }
          });
        }}
        didUpdate={(
          { options: prevOptions },
          { options: nextOptions, state: { marker } },
        ) => {
          const options = pickChangedProps(prevOptions, nextOptions);

          if (options) {
            marker.setOptions(options as google.maps.MarkerOptions);
          }
        }}
        willUnmount={({ maps, state: { marker } }) => {
          marker.setMap(null);
          maps.event.clearInstanceListeners(marker);
        }}
        render={({ state: { ctx } }) =>
          typeof icon === "string" ? null : (
            <MarkerContextProvider value={ctx}>{icon}</MarkerContextProvider>
          )
        }
      />
    );
  }
}
