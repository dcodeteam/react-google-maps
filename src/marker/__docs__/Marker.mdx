---
name: Marker
---

import { Playground, PropsTable } from 'docz'

import { Marker } from '../Marker'
import { GoogleMap } from '../../google-map/GoogleMap'

# GoogleMap

Draws `google.maps.Marker`.

**Google Maps Docs:**
* [google.maps.Marker](https://developers.google.com/maps/documentation/javascript/reference#Marker)
* [google.maps.MarkerOptions](https://developers.google.com/maps/documentation/javascript/reference#MarkerOptions)

<Playground>
  <GoogleMap
    zoom={8}
    maps={google.maps}
    style={{ height: '320px' }}
    center={{ lat: 36.964, lng: -122.015 }}
  >
    <Marker
      animation="BOUNCE"
      draggable={true}
      position={{ lat: 36.964, lng: -122.015 }}
      onDragEnd={() => {
        alert("It will go back")
      }}
    />
  </GoogleMap>
</Playground>

<PropsTable of={Marker} />
