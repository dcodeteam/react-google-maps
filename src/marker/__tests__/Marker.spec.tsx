import { mount } from "enzyme";
import * as React from "react";

import { mockMarker } from "../../__tests__/testUtils";
import { GoogleMapContextProvider } from "../../google-map-context/GoogleMapContext";
import { Marker, MarkerProps } from "../Marker";
import { MarkerContextConsumer } from "../MarkerContext";
import { MarkerEvent } from "../MarkerEvent";

const map = new google.maps.Map(null);

function MockMarker(props: MarkerProps) {
  return (
    <GoogleMapContextProvider value={{ map, maps: google.maps }}>
      <Marker {...props} />
    </GoogleMapContextProvider>
  );
}

describe("Marker", () => {
  beforeEach(() => {
    jest.restoreAllMocks();
  });

  it("should create marker and attach it to map on mount", () => {
    const markerSpy = mockMarker();

    mount(<MockMarker position={{ lat: 0, lng: 1 }} />);

    expect(markerSpy).toBeCalledTimes(1);

    const [{ value: marker }] = markerSpy.mock.results;

    expect(marker.setMap).toBeCalledTimes(1);
    expect(marker.setMap).toHaveBeenLastCalledWith(map);
  });

  it("should set default options on mount", () => {
    const markerSpy = mockMarker();

    mount(<MockMarker position={{ lat: 0, lng: 1 }} />);

    const [{ value: marker }] = markerSpy.mock.results;

    expect(marker.setOptions.mock.calls).toMatchSnapshot();
  });

  it("should set custom options on mount", () => {
    const markerSpy = mockMarker();

    mount(
      <MockMarker
        animation="BOUNCE"
        clickable={true}
        cursor="pointer"
        draggable={true}
        icon="https://url.to/icon.png"
        label="A"
        opacity={0.5}
        optimized={false}
        position={{
          lat: 0,
          lng: 1
        }}
        shape={{
          type: "foo"
        }}
        title="Foo"
        zIndex={1000}
        visible={false}
      />
    );

    const [{ value: marker }] = markerSpy.mock.results;

    expect(marker.setOptions.mock.calls).toMatchSnapshot();
  });

  it("should add listeners without handlers", () => {
    const markerSpy = mockMarker();

    mount(<MockMarker position={{ lat: 0, lng: 1 }} />);

    const [{ value: marker }] = markerSpy.mock.results;

    expect(marker.addListener.mock.calls).toMatchSnapshot();
  });

  it("should add listeners with handlers", () => {
    const markerSpy = mockMarker();
    const props: MarkerProps = { position: { lat: 0, lng: 1 } };
    const keys = Object.keys(MarkerEvent) as Array<keyof MarkerProps>;

    keys.forEach(key => {
      props[key] = jest.fn();
    });

    mount(<MockMarker {...props} />);

    const [{ value: marker }] = markerSpy.mock.results;

    keys.forEach(key => {
      const event = MarkerEvent[key as keyof typeof MarkerEvent];
      const handler = props[key];
      const payload = { key, event };

      expect(handler).toBeCalledTimes(0);

      marker.emit(event, payload);

      expect(handler).toBeCalledTimes(1);
      expect(handler).lastCalledWith(payload);
    });
  });

  it("should render icon if its valid react element", () => {
    const wrapper = mount(
      <MockMarker position={{ lat: 0, lng: 1 }} icon={<div>Foo</div>} />
    );

    const divWrapper = wrapper.find("div");

    expect(divWrapper.length).toBe(1);
    expect(divWrapper.html()).toBe("<div>Foo</div>");
  });

  it("should pass `MarkerContext` to children", () => {
    const consumer = jest.fn();
    const markerSpy = mockMarker();

    mount(
      <MockMarker
        position={{ lat: 0, lng: 1 }}
        icon={<MarkerContextConsumer>{consumer}</MarkerContextConsumer>}
      />
    );

    const [{ value: marker }] = markerSpy.mock.results;

    expect(consumer).toBeCalledTimes(1);
    expect(consumer).lastCalledWith({ marker });
  });

  it("should reset position on drag end", () => {
    const onDragEnd = jest.fn();
    const markerSpy = mockMarker();
    const position = { lat: 0, lng: 1 };

    mount(<MockMarker position={position} onDragEnd={onDragEnd} />);

    const [{ value: marker }] = markerSpy.mock.results;

    expect(onDragEnd).toHaveBeenCalledTimes(0);
    expect(marker.setPosition).toHaveBeenCalledTimes(0);

    marker.emit(MarkerEvent.onDragEnd);

    expect(onDragEnd).toHaveBeenCalledTimes(1);
    expect(marker.setPosition).toHaveBeenCalledTimes(1);
    expect(marker.setPosition).toHaveBeenLastCalledWith(position);
  });

  it("should update only changed options on props update", () => {
    const markerSpy = mockMarker();
    const wrapper = mount(<MockMarker position={{ lat: 0, lng: 1 }} />);

    const [{ value: marker }] = markerSpy.mock.results;

    expect(marker.setOptions).toBeCalledTimes(1);

    wrapper.setProps({ clickable: true });

    expect(marker.setOptions).toBeCalledTimes(2);
    expect(marker.setOptions).lastCalledWith({ clickable: true });

    wrapper.setProps({ clickable: true });

    expect(marker.setOptions).toBeCalledTimes(2);

    wrapper.setProps({ clickable: false });

    expect(marker.setOptions).toBeCalledTimes(3);
    expect(marker.setOptions).lastCalledWith({ clickable: false });
  });

  it("should remove from map on unmount", () => {
    const markerSpy = mockMarker();
    const wrapper = mount(<MockMarker position={{ lat: 0, lng: 1 }} />);

    const [{ value: marker }] = markerSpy.mock.results;

    expect(marker.setMap).toBeCalledTimes(1);

    wrapper.unmount();

    expect(marker.setMap).toBeCalledTimes(2);
    expect(marker.setMap).lastCalledWith(null);

    expect(google.maps.event.clearInstanceListeners).toBeCalledTimes(1);
    expect(google.maps.event.clearInstanceListeners).lastCalledWith(marker);
  });
});
