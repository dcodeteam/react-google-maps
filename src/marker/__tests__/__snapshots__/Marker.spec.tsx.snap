// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker should add listeners without handlers 1`] = `
Array [
  Array [
    "dragend",
    [Function],
  ],
  Array [
    "click",
    [Function],
  ],
  Array [
    "dblclick",
    [Function],
  ],
  Array [
    "rightclick",
    [Function],
  ],
  Array [
    "mouseout",
    [Function],
  ],
  Array [
    "mouseover",
    [Function],
  ],
  Array [
    "mousedown",
    [Function],
  ],
  Array [
    "mouseup",
    [Function],
  ],
  Array [
    "drag",
    [Function],
  ],
  Array [
    "dragstart",
    [Function],
  ],
  Array [
    "dragend",
    [Function],
  ],
  Array [
    "position_changed",
    [Function],
  ],
]
`;

exports[`Marker should set custom options on mount 1`] = `
Array [
  Array [
    Object {
      "animation": "BOUNCE",
      "clickable": true,
      "cursor": "pointer",
      "draggable": true,
      "icon": "https://url.to/icon.png",
      "label": "A",
      "opacity": 0.5,
      "optimized": false,
      "position": Object {
        "lat": 0,
        "lng": 1,
      },
      "shape": Object {
        "type": "foo",
      },
      "title": "Foo",
      "visible": false,
      "zIndex": 1000,
    },
  ],
]
`;

exports[`Marker should set default options on mount 1`] = `
Array [
  Array [
    Object {
      "animation": undefined,
      "clickable": undefined,
      "cursor": undefined,
      "draggable": undefined,
      "icon": undefined,
      "label": undefined,
      "opacity": undefined,
      "optimized": undefined,
      "position": Object {
        "lat": 0,
        "lng": 1,
      },
      "shape": undefined,
      "title": undefined,
      "visible": undefined,
      "zIndex": undefined,
    },
  ],
]
`;
